### ISSUE TITLE
{"format":"uri"} doesn't seem to be used

### ISSUE THREAD

**Issue by skandragon** [Julian/jsonschema#713]
I have a schema defined as:

```
    submit_log_schema = {
        "type": "object",
        "required": ["v1Location", "v2Location", "resultPrefix", "algorithm"],
        "additionalProperties": False,
        "properties": {
            "v1Location": {"type": "string", "format": "uri"},
            "v2Location": {"type": "string", "format": "uri"},
            "resultPrefix": {"type": "string", "format": "uri"},
            "algorithm": {"type": "string", "enum": ["spell"]}
        }
    }
```

When I validate against this, if one of the 'uri' format strings is not a URI, it still passes validation.

Examples:  empty string, or '-123' -- either the URI spec is so loose I'm not coming up with examples, or I have something configured incorrectly.

I do have rfc3987 installed.  I'm using jsonschema 3.2.0 (via `pip install jsonschema[format]`)

**Comment by Julian** [comment]
Hi. Hard to tell without seeing what you're running.

Did you follow the [validating formats](https://python-jsonschema.readthedocs.io/en/stable/validate/#validating-formats) docs? Specifically, are you forgetting to pass in a format checker?

**Comment by skandragon** [comment]
For better or worse, here's the class. The intent is to hide the internals and just return a nice list of errors if any.

```class V1ApiSchema:
    submit_log_schema = {
        "type": "object",
        "required": ["v1Location", "v2Location", "resultPrefix", "algorithm"],
        "additionalProperties": False,
        "properties": {
            "v1Location": {"type": "string", "format": "uri"},
            "v2Location": {"type": "string", "format": "uri"},
            "resultPrefix": {"type": "string", "format": "uri"},
            "algorithm": {"type": "string", "enum": ["spell"]}
        }
    }

    def __init__(self):
        self.submit_log_validator = Draft7Validator(self.submit_log_schema)

    def validate_submit_log(self, instance: dict) -> Union[None, List[dict]]:
        errors = []
        for error in sorted(self.submit_log_validator.iter_errors(instance), key=str):
            errors.append({ "path": list(error.path), "message": error.message})
        if len(errors) > 0:
            return errors
        return None
```

**Comment by Julian** [comment]
Thanks. Looks like the code there indeed isn't passing in a format checker.

Have a look at the docs I linked, they should get you sorted.

**Comment by skandragon** [comment]
Hmm, one would assume that if I'm using a `Draft7Validator` object it would be configured to handle this.

**Comment by Julian** [comment]
Not to be *too* snarky but... one would assume when presented not once but twice with links to the docs (which cover this exact issue twice, and explain why the behavior is the way it is -- which is mandated by the specification you are using), that you'd have a look first, and that hopefully after doing so, you'd understand what you need to do and that'd be the end of it.

I'm sorry you had difficulty, and it's not uncommon.

**Comment by skandragon** [comment]
I'm sorry, I did read the docs.  What I'm describing is more what I would expect.  I understand when using the generic `validate()` method I would need to supply all the details on what I want to have applied for validation, but when using the more specific `Draft7Validator` methods, I would expect it to be more configured out of the box as per draft 7 features.

I admit I am not well versed in the json validation world, I'm just expressing what I expected vs what happened.  I did read the docs, and I did find the solution (which is now working) and I thank you for that.  I'm also sorry I did not research the draft mandates as fully as I could have.  From my cursory read of them, unknown formats was mandated to be ignored (that is, the rules not applied) but I thought the uri format was "known" in this context, and would just work.

Now I know better, and my code works, so thank you for the pointer.
